lib/internal/modules/cjs/loader.js

```diff
@@ -679,19 +716,28 @@ Module.prototype._compile = function(content, filename) {
 
   content = stripShebang(content);
 
-  // create wrapper function
-  var wrapper = Module.wrap(content);
-
-  var compiledWrapper = vm.runInThisContext(wrapper, {
-    filename: filename,
-    lineOffset: 0,
-    displayErrors: true,
-    importModuleDynamically: experimentalModules ? async (specifier) => {
-      if (asyncESM === undefined) lazyLoadESM();
-      const loader = await asyncESM.loaderPromise;
-      return loader.import(specifier, normalizeReferrerURL(filename));
-    } : undefined,
-  });
+  let compiledWrapper;
+  if (patched) {
+    const wrapper = Module._wrap(content);
+    compiledWrapper = vm.runInThisContext(wrapper, {
+      filename,
+      lineOffset: 0,
+      displayErrors: true,
+      importModuleDynamically: experimentalModules ? async (specifier) => {
+        if (asyncESM === undefined) lazyLoadESM();
+        const loader = await asyncESM.loaderPromise;
+        return loader.import(specifier, normalizeReferrerURL(filename));
+      } : undefined,
+    });
+  } else {
+    compiledWrapper = vm.compileFunction(content, [
```

Comment:
I'd prefer we use `internalBinding('contextify').compileFunction` instead of this - for one we don't need the argument checking overhead in the public API here, and another users can also monkey patch `vm.compileFunction` then we'll be repeating history

Reply:
Great idea, I've gone ahead and implemented this.

@devsnek I've added a test that `vm.compileFunction` does not support dynamic import at all now. So this is only an internal usage. So this PR now has no effect on the vm API for users. Does that mitigate your concerns here?